### ISSUE TITLE
Add buildship auto sync preference

### ISSUE THREAD

**Issue by Frank-D** [redhat-developer/vscode-java#2673]
hello everyone, 
**TLDR**: anyone knows how to control the `.settings/org.eclipse.buildship.core.prefs` > `auto.sync` flag in VSC, without touching that file specifically? 
(...to make it 'true', as it was false by default for me and at the origin of my problem explained below..) 
*************************

I'm working on a *gradle / java / spring-boot* project using vsc (with '*extension pack for java*' + '*gradle for java*' extensions installed to name a few..), and I've been losing hours trying to figure out how updating my `build.gradle` file by adding/removing dependencies can automatically update my vsc java coding context so that I can have access to those libs in my code, more precisely, how can the `Java Projects > ... > Project and External Dependencies` UI get updated automatically when I update and save new dependencies in my `build.gradle` file...

After hours of research, I managed to find the setting `java.import.generatesMetadataFilesAtProjectRoot` (from this [link](https://github.com/eclipse/eclipse.jdt.ls/pull/1900)) , I turned it on, and then discovered the new generated file `.settings/org.eclipse.buildship.core.prefs` in my project root.

One of its setting captured my attention, e.g. `auto.sync=false`, I wondered what if I turn it on and try just for fun, and boom, it solved my problem!

Now, with `.settings/org.eclipse.buildship.core.prefs` > `auto.sync=true`, I can finally update my `build.gradle` file, add/remove dependencies, save the file, and finally (actually, even if I don't answer "Yes" to the notification asking if I want to synchronize my java classpath/configuration and let the notif. displayed for a little while..), I will see the dependencies changes being reflected in vs code, e.g. changes being reflected under `Java Projects` tab of the explorer under its `Project and External Dependencies`, which directly means, I can finally use those libs in my java code (before that, I couldn't, the libs were not recognized / could not be imported by vs code, and my only way around was to do a `Java > Clean Java Language Server Workspace`, and get vs code restarted, which was super annoying for every single gradle dependency that I was playing with, lots of time lost restarting vsc all the time, generating frustration as usual etc..).

Now my simple question is, what is the "official" way of getting the `auto.sync` flag to be `true`, without having to manually modify that `.settings/org.eclipse.buildship.core.prefs` file myself? As I understand it, we are not supposed to play with that file directly, as it gets generated by the Java Language Server extension, based on some other external settings/inputs?

Any help much appreciated, thanks!

##### Environment
- Operating System: macOS big sur (11.6.5)
- JDK version: openjdk version "17.0.4.1" 2022-08-12
- Visual Studio Code version: 1.71.0
- Java extension version: 1.10.0

##### Steps To Reproduce
1. **Install** `VS Code` and at least the following extensions: `Extension pack for Java` + `Gradle for Java`.
2. **Create** a new simple gradle java project (the `gradle for java` extension can help you with that, e.g. use the vsc command `Gradle: Create a Gradle Java Project... (Advanced)`.
3. **Add** new gradle dependencies in your `build.gradle` file, and **Save** the file changes.
4. When asked by a vsc notification if you want to synchronize the java classpath/configuration, say **Yes**.
5. **Go** to `Explorer` > `Java Projects` > `[...your app module...]` > `Project and External Dependencies`, and **expend** the list.

Here's a simple `build.gradle` file that I used to reproduce the problem described above (but the exact same file will work with that `auto.sync=true` though..):
```
plugins {
    // Apply the application plugin to add support for building a CLI application in Java.
    id 'application'
    // Apply the java plugin
    id 'java'
}

sourceCompatibility = '17'

repositories {
    // Use Maven Central for resolving dependencies.
    mavenCentral()
}

dependencies {
    implementation 'org.apache.commons:commons-lang3:3.0'
}
    
application {
    // Define the main class for the application.
    mainClass = 'com.demo.App'
}
```

##### Current Result
Notice that the gradle dependencies added at `step 3` are not listed under `step 5`, e.g. Gradle and VSC (vscode-java e.g. the Java Language Server) remain out of sync in terms of dependencies.
Consequence of the above is that those dependencies/libs won't be available in my java code, so I'm unable to import any of them or else I'll get java compiling errors.

##### Expected Result
Gradle and VSC/vscode-java have my dependencies automatically synchronized.
Consequence of the above is that those dependencies/libs BECOME automatically available in my java code, so I'm able to import any of them without any java compiling errors.

##### Additional Informations
One workaround to fix the described situation and ensure `step 5` above would list the added dependencies of `step 3` above, was to execute the vsc command `Java > Clean Java Language Server Workspace`, but that is super annoying to do, every single time you have to modify your gradle dependencies (add/remove), like in `step 3` above, since it requires you to restart your entire editor/vsc, and it then takes a little while for it to be back opened and ready, since the java language server will take a little while to reload, re-index everything etc. But at that point, it would at least be finally in sync with my `build.gradle` file. 

Obviously, the other solution that I found around the `.settings/org.eclipse.buildship.core.prefs` > `auto.sync=true` is much much better and actually fixes everything for me, but I'm looking for the right way to fix this through any other vs code settings, rather than modifying directly that file myself, if at all possible.

**Comment by Frank-D** [comment]
hi @snjeza and thanks for supporting me/us on this one!

Just to make sure, will your solution also solve projects where the internal java setting files are still stored inside the internal vscode workspaceStorage location? (e.g. setting `java.import.generatesMetadataFilesAtProjectRoot=false`, e.g. default value)

I've another project where I had the exact same problem as I originally described in this ticket, however for this project I decided to keep the internal settings into the workspaceStorage location instead of having all those extra internal files at the root of my project and visible.

In order to fix the auto-sync problem between gradle and vscode-java, I needed to apply the manual fix as originally described, but obviously not at the root of my project (<project root>/.settings/org.eclipse.buildship.core.prefs), but instead, to the below location (I'm on macOS):

```
# pwd                                                                                                                                 
/Users/<my_user>/Library/Application Support/Code/User/workspaceStorage/<my_current_workspace_storage_id>/redhat.java

# grep -r "auto.sync" .
./jdt_ws/.metadata/.plugins/org.eclipse.core.resources/.projects/pipeline-core/.settings/org.eclipse.buildship.core.prefs:auto.sync=false
```

For the record, for people like me having a multi-modules project, for example multiple gradle submodules B, C and D at the root of a main project A, there will be multiple directories located at the location specified above, one for each submodule B, C and D, but also, one for the overall project A. 

Each will have a file "./jdt_ws/.metadata/.plugins/org.eclipse.core.resources/.projects/`<project_or_submodule_name>`/.settings/org.eclipse.buildship.core.prefs", ***BUT ONLY*** the file of the main project A should be modified to have the "auto.sync=true" setting value; the other submodules don't need to have this setting set for the overall fix to work.

@snjeza, will your patch also fix/account for the above situation? 
E.g. a multi-modules gradle project with the vscode-java internal setting/config files stored in the default location -> in the workspaceStorage..

Thanks

**Comment by snjeza** [comment]
@Frank-D you can check https://github.com/snjeza/vscode-test/raw/master/java-1.10.1.vsix
You have to add 
```
"java.import.gradle.auto.sync": true
```
to settings.json

**Comment by Frank-D** [comment]
**Tested, working perfectly well.**

No need to play with the internal `org.eclipse.buildship.core.prefs` file anymore, since the new workspace settings.json `"java.import.gradle.auto.sync": true` introduced by the above fix will do it just fine.

It is working for all the scenarios that I've mentioned above, tested. 
(e.g., working with internal metadata files at project root OR in workspaceStorage (e.g. `"java.import.generatesMetadataFilesAtProjectRoot": true | false`), and working for gradle java `single module` project OR gradle java `multi-modules` project).

**Thanks @snjeza , great job and super fast response and fix !**
(took you only 30min to respond to my [original msg in gitter](https://gitter.im/redhat-developer/vscode-java?at=63209bdaf4d7a323deceef2b), and between the time I reported this issue and the time you came back with a working fix, 3 hours, so all that within the same day, super impressive response, keep on the great work!)

**Comment by snjeza** [comment]
It hasn't been reviewed and merged yet.

**Comment by Frank-D** [comment]
@snjeza, I've currently having VS Code `1.71.0` installed, with multiple workspaces configured for different projects, so I wasn't really planning on messing up with those existing vsc configs/settings and re-install an earlier version of the entire editor, especially that you're asking vs code `0.10.0`, this confuses me since I can't even find such an old version(!), the earliest I could find was [0.10.3](https://code.visualstudio.com/updates/vNovember) and it's a beta version from back 2015, that's 7+ years ago! Have you made a typo?

Having said that, I've always had the setting that you're talking about
```
"java.configuration.updateBuildConfiguration": "automatic",
```
applied in my vsc workspace, and still, I was not getting the gradle and java language server extensions to communicate well together, until we turned on that said `auto.sync=true`.

One thing I've noticed in my vsc java logs though, is that I'm getting in both the `client.log.<today's date>` and the server `.log` log files the following error, and it is still happening as of today, could it have any link with blocking the dependencies synchronization between gradle and the java language server (..and hence, forcing me to have that `auto.sync=true` set) ?

```
!ENTRY org.eclipse.buildship.core 2 0 2022-09-13 18:21:33.663
!MESSAGE Cannot collect dependency scope information for launch configuration 1663107693658
!STACK 1
org.eclipse.core.runtime.CoreException: The input type of the launch configuration does not exist
	at org.eclipse.jdt.junit.launcher.JUnitLaunchConfigurationDelegate.abort(JUnitLaunchConfigurationDelegate.java:774)
	at org.eclipse.jdt.junit.launcher.JUnitLaunchConfigurationDelegate.getTestTarget(JUnitLaunchConfigurationDelegate.java:768)
	at org.eclipse.jdt.junit.launcher.JUnitLaunchConfigurationDelegate.evaluateTests(JUnitLaunchConfigurationDelegate.java:369)
	at org.eclipse.buildship.core.internal.launch.SupportedLaunchConfigType$RelaxedJUnitLaunchConfigurationDelegate.evaluateTests(SupportedLaunchConfigType.java:139)
	at org.eclipse.buildship.core.internal.launch.SupportedLaunchConfigType$2.getSourceFolders(SupportedLaunchConfigType.java:70)
	at org.eclipse.buildship.core.internal.launch.SupportedLaunchConfigType.collectSourceFolders(SupportedLaunchConfigType.java:110)
	at org.eclipse.buildship.core.internal.launch.LaunchConfigurationScope.from(LaunchConfigurationScope.java:57)
	at org.eclipse.buildship.core.internal.workspace.GradleClasspathContainerRuntimeClasspathEntryResolver.resolveRuntimeClasspathEntry(GradleClasspathContainerRuntimeClasspathEntryResolver.java:55)
	at org.eclipse.jdt.internal.launching.RuntimeClasspathEntryResolver.resolveRuntimeClasspathEntry(RuntimeClasspathEntryResolver.java:49)
	at org.eclipse.jdt.launching.JavaRuntime.resolveRuntimeClasspathEntry(JavaRuntime.java:1265)
	at org.eclipse.buildship.core.internal.launch.GradleClasspathProvider.resolveClasspath(GradleClasspathProvider.java:101)
	at org.eclipse.jdt.internal.launching.RuntimeClasspathProvider.resolveClasspath(RuntimeClasspathProvider.java:68)
	at org.eclipse.jdt.launching.JavaRuntime.resolveRuntimeClasspath(JavaRuntime.java:1667)
	at org.eclipse.jdt.launching.AbstractJavaLaunchConfigurationDelegate.getClasspathAndModulepath(AbstractJavaLaunchConfigurationDelegate.java:464)
	at org.eclipse.jdt.ls.core.internal.commands.ProjectCommand$1.run(ProjectCommand.java:173)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2380)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2405)
	at org.eclipse.jdt.ls.core.internal.commands.ProjectCommand.getClasspaths(ProjectCommand.java:170)
	at org.eclipse.jdt.ls.core.internal.JDTDelegateCommandHandler.executeCommand(JDTDelegateCommandHandler.java:82)
	at org.eclipse.jdt.ls.core.internal.handlers.WorkspaceExecuteCommandHandler$1.run(WorkspaceExecuteCommandHandler.java:230)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.jdt.ls.core.internal.handlers.WorkspaceExecuteCommandHandler.executeCommand(WorkspaceExecuteCommandHandler.java:220)
	at org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.lambda$4(JDTLanguageServer.java:529)
	at org.eclipse.jdt.ls.core.internal.BaseJDTLanguageServer.lambda$0(BaseJDTLanguageServer.java:75)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:483)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
!SUBENTRY 1 org.eclipse.jdt.junit.core 4 101 2022-09-13 18:21:33.663
!MESSAGE The input type of the launch configuration does not exist
```

**Comment by snjeza** [comment]
>  vs code 0.10.0,

I'm sorry. I thought 1.10.0

**Comment by snjeza** [comment]
> Having said that, I've always had the setting that you're talking about
> "java.configuration.updateBuildConfiguration": "automatic",


Could you check your <project_root>/.vscode/settings.json ?

**Comment by snjeza** [comment]
> ... (actually, even if I don't answer "Yes" to the notification asking if I want to synchronize...

This notification will be shown if you set
```
"java.configuration.updateBuildConfiguration": "interactive",
```
probably in  <project_root>/.vscode/settings.json

**Comment by Frank-D** [comment]
hey @snjeza, re
> Could you check your <project_root>/.vscode/settings.json ?

I've tried both "interactive" (with the vsc notification where I was answering Yes to the question if I wanted to sync the java classpath/configuration after updating my build.gradle file..) and "automatic", but in any case, same result, my java projects dependencies were not getting updated with that "java.configuration.updateBuildConfiguration" setting alone..

Here'e my .vscode/settings.json right now (which includes the temp fix "java.import.gradle.auto.sync" that you shared with me above, but before adding it to ~temporary fix the sync issue (until the PR gets accepted & merged..), I had all the other settings below except that one, such as "java.configuration.updateBuildConfiguration": "automatic", and still, it was not syncing at all between gradle and java lang server when modifying my gradle build files...):
```
{
    "gradle.focusTaskInExplorer": true,
    "gradle.javaDebug.cleanOutput": true,
    "gradle.debug": true,
    "gradle.disableConfirmations": true,
    "java.autobuild.enabled": true,
    "java.dependency.syncWithFolderExplorer": true,
    "java.configuration.updateBuildConfiguration": "automatic",
    "java.import.gradle.enabled": true,
    "java.import.gradle.wrapper.enabled": true,
    "java.import.gradle.auto.sync": true,
    ...
    ...some more gradle/java-unrelated personal vsc settings...
    ...
}

```

**Comment by Frank-D** [comment]
Now, I see that your PR was updated and merged, with a few comments regarding merging the temporary-introduced "java.import.gradle.auto.sync" setting into "java.configuration.updateBuildConfiguration" (which I think is a good idea, as long as we're not introducing any bad side effects..), so was this the final change that you made and which got merged?

**Comment by snjeza** [comment]
> was this the final change that you made and which got merged?

@Frank-D Yes, it was.
I have tested it and works fine. If you set

```
 "java.configuration.updateBuildConfiguration": "automatic",
```
jdtls will add `auto.sync=true` to your .settings/org.eclipse.buildship.core.prefs
You can try https://github.com/snjeza/vscode-test/raw/master/java-1.10.3.vsix

**Comment by Frank-D** [comment]
Now, I see that your PR was updated and merged, with a few comments regarding merging the temporary-introduced "java.import.gradle.auto.sync" setting into "java.configuration.updateBuildConfiguration" (which I think is a good idea, as long as we're not introducing any bad side effects..), so was this the final change that you made and which got merged?

Looking at the merged PR changes, I see 2 things for which I'm not exactly sure I understand why as I don't see any code comment/explanation for it, for instance:

1- https://github.com/eclipse/eclipse.jdt.ls/commit/64740136c8c78563c0e55e8a464b04bd218495c0#diff-f23d0258ce24117c110acd6753e8490a09fbdf0bc039d662bdffe279c8c28469L241-R246

I see that you are now skipping the "updateProject(...)" call if the project is gradle, why is that exactly, what does that mean? I see that you've added a reference to current ticket as a ~code comment for this skipping behaviour, but to me it still does explain exactly why we are skipping that method call if gradle, could you explain why please?

2- https://github.com/eclipse/eclipse.jdt.ls/commit/64740136c8c78563c0e55e8a464b04bd218495c0#diff-4be4f273e609fde1220c2d89f532f1fe207f280c100126555c3747fcf44d18d8R448
and
https://github.com/eclipse/eclipse.jdt.ls/commit/64740136c8c78563c0e55e8a464b04bd218495c0#diff-f23d0258ce24117c110acd6753e8490a09fbdf0bc039d662bdffe279c8c28469R241

I see that in both code change areas, the implemented fix seems to only apply IF we have the `"java.configuration.updateBuildConfiguration"` set to `"automatic"` right? But what if someone wants to use the "interactive" behaviour instead, I guess your fix won't apply then right? So a user would receive the vsc notification asking if they want to synchronize their java classpath/configuration, they would answer Yes, but then nothings would happen?

On the other hand, as I explained in one of my previous message up there, it seems like the eclipse buildship lib doesn't care about the java language server right (? ..probably since only the java language server knows about buildship since it's a wrapper around or it just makes use of it and not the other way around..), and so, with your previous fix where I was able to control the eclipse buildship's `auto.sync` flag by using your temporary introduced `"java.import.gradle.auto.sync": true` solution, it did solve my gradle->java lang server dependencies sync issue like I already said a few times before, but that is, completely independently from the `"java.configuration.updateBuildConfiguration"` setting (I guess that one could have been set to `"disabled"` and it would still have sync'ed properly. 

The reason why I'm saying that, is that I tried setting it at `"interactive"` at some point to have more control over the timing of things, so I was receiving the vsc notification upon saving my build.gradle file (say after adding or removing a dependency..), but then even without answering Yes (..or not) to the notification java classpath/configuration sync question and waiting instead a few seconds with the notification displayed, my `Java Projects`'s app dependencies were getting sync'ed properly anyways(!), because of that `"java.import.gradle.auto.sync": true` setting (..which under the hood set eclipse buildship's auto.sync flag to true, the real fix for me actually..). 

Anyways, I'll let you comment at this point, but I just wanted to point that out, that 1- the ultimate fix to all that should work IMO for both `"java.configuration.updateBuildConfiguration"` set to `"automatic"` AND `"interactive"` (that is, if user then answers YES to the notification..), PLUS I wanted to highlight that in the previous fix that you shared with me, unless that logic has changed in your last PR, the 'updateBuildConfiguration' set to "automatic" (or "interactive" + answering Yes to the notification) had nothing to do with the synchronization happening behind the scene as soon as I was saving my build.gradle file, thanks to that `"java.import.gradle.auto.sync": true` temp setting (aka `auto.sync = true`).

**Comment by Frank-D** [comment]
> > was this the final change that you made and which got merged?
> 
> @Frank-D Yes, it was. I have tested it and works fine. If you set
> 
> ```
>  "java.configuration.updateBuildConfiguration": "automatic",
> ```
> 
> jdtls will add `auto.sync=true` to your .settings/org.eclipse.buildship.core.prefs You can try https://github.com/snjeza/vscode-test/raw/master/java-1.10.3.vsix

OK, I'll give it a shot thanks for that, but feel free to read my previous comment (sry it's long..), as I have a few doubts on the solution in some areas.. thanks

**Comment by snjeza** [comment]
> I see that you are now skipping the "updateProject(...)" call if the project is gradle, why is that exactly, what does that mean? I see that you've added a reference to current ticket as a ~code comment for this skipping behaviour, but to me it still does explain exactly why we are skipping that method call if gradle, could you explain why please?

It works in the following way:
```
 "java.configuration.updateBuildConfiguration": "automatic",
```

javals set `auto.sync=true`, buildship synchronizes the projects,  the `updateProject(...` call is skipped

```
 "java.configuration.updateBuildConfiguration": "interactive",
```

javals set `auto.sync=false`, you control sync using the notification.

```
 "java.configuration.updateBuildConfiguration": "disabled",
```

nobody synchronizes the projects. You can do it using the `Reload Projects` command.
